chronos/asyncmacro2.nim

```diff
@@ -247,6 +246,19 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
     closureIterator.pragma = newNimNode(nnkPragma, lineInfoFrom=prc.body)
     closureIterator.addPragma(newIdentNode("closure"))
 
+    # TODO when push raises is active in a module, the iterator here inherits
```

Comment:
It should probably inherit the exceptions from the `{.async.}` proc? Should be a matter of reading them from the annotated proc and attaching to the iterator?

Reply:
it needs to either a) deduce exceptions which it can't do easily, or b) use explicit annotations which come with their own issues - once it knows what exceptions are raised, it needs to rewrite all `await` calls with some additional macro logic - this is far beyond this PR to solve.

Reply:
Fair, I think (a) is probably closer to what is needed and it should be possible to read annotations from the wrapped proc, this will happen on macro expansion at compile time so the original exception info should still be present on the wrapped proc/method?

Reply:
the issue is `await` that in its present form always causes a `CatchableError` raises effect because it reads from the future which doesn't hold specific exception type - one would need to transport the error type more precisely in some shape or form - there are a couple proposals for that - the main point of _this_ branch is to get rid of `Exception` effects that leak out and also address the `poll` exception leaks once and for all.

Reply:
OK, what I mean is that the _desirable_ option is probably (a) and it's fine we address it as part of another PR.